system/lib/wasmfs/file_table.cpp

```diff
@@ -63,12 +75,35 @@ class StderrFile : public File {
   __wasi_errno_t read(const uint8_t* buf, __wasi_size_t len) override {
     return __WASI_ERRNO_INVAL;
   };
+
+public:
+  static std::shared_ptr<StderrFile> getSharedPtr() {
+    static const std::shared_ptr<StderrFile> stderrFile = std::make_shared<StderrFile>();
+    return stderrFile;
+  }
 };
 
 FileTable::FileTable() {
-  entries.push_back(std::make_shared<OpenFileState>(0, std::make_shared<StdinFile>()));
-  entries.push_back(std::make_shared<OpenFileState>(0, std::make_shared<StdoutFile>()));
-  entries.push_back(std::make_shared<OpenFileState>(0, std::make_shared<StderrFile>()));
+  entries.push_back(std::make_shared<OpenFileState>(0, StdinFile::getSharedPtr()));
+  entries.push_back(std::make_shared<OpenFileState>(0, StdoutFile::getSharedPtr()));
+  entries.push_back(std::make_shared<OpenFileState>(0, StderrFile::getSharedPtr()));
+}
+
+RootDirectory::RootDirectory() {
+  // Avoid calling the destructor -
+  // go/totw/110#tip-of-the-week-110-safe-initialization-of-global-and-static-objects
+
+  // Obtain other dev/stdin, dev/stdout members, ensuring that they are initialized on demand
+  static const std::shared_ptr<Directory> devDirectory = [] {
+    return std::make_shared<Directory>();
+  }();
```

Comment:
Is `static` needed here? I think just creating a shared ptr, and then saving it on `entries` would keep it alive.

Or is the only reason for `static` that it avoids calling the destructor?

Reply:
I don't think the lambda should be necessary here, either.

Reply:
True, I agree the `static` is unnecessary here.